nibabel/streamlines/tck.py

```diff
@@ -0,0 +1,444 @@
+                      for k, v in header.items()
+                      if k not in exclude and not k.startswith("_")])
+        lines.append("file: . ")  # Manually add this last field.
+        out = "\n".join(lines)
+
+        # Check the header is well formatted.
+        if out.count("\n") > len(lines) - 1:  # \n only allowed between lines.
+            msg = "Key-value pairs cannot contain '\\n':\n{}".format(out)
+            raise HeaderError(msg)
+
+        if out.count(":") > len(lines):  # : only one per line.
+            msg = "Key-value pairs cannot contain ':':\n{}".format(out)
+            raise HeaderError(msg)
+
+        # Write header to file.
+        fileobj.write(asbytes(out))
+
+        hdr_len_no_offset = len(out) + 5
+        # Need to add number of bytes to store offset as decimal string. We
+        # start with estimate without string, then update if the
+        # offset-as-decimal-string got longer after adding length of the
+        # offset string.
+        new_offset = -1
+        old_offset = hdr_len_no_offset
+        while new_offset != old_offset:
+            old_offset = new_offset
+            new_offset = hdr_len_no_offset + len(str(old_offset))
+
+        fileobj.write(asbytes(str(new_offset) + "\n"))
+        fileobj.write(asbytes("END\n"))
+
+    @staticmethod
+    def _read_header(fileobj):
+        """ Reads a TCK header from a file.
+
+        Parameters
+        ----------
+        fileobj : string or file-like object
+            If string, a filename; otherwise an open file-like object
+            pointing to TCK file (and ready to read from the beginning
+            of the TCK header). Note that calling this function
+            does not change the file position.
+
+        Returns
+        -------
+        header : dict
+            Metadata associated with this tractogram file.
+        """
+        # Record start position if this is a file-like object
+        start_position = fileobj.tell() if hasattr(fileobj, 'tell') else None
+
+        with Opener(fileobj) as f:
+            # Read magic number
+            magic_number = asstr(f.fobj.readline()).strip()
+
+            # Read all key-value pairs contained in the header.
+            buf = asstr(f.fobj.readline())
+            while not buf.rstrip().endswith("END"):
+                buf += asstr(f.fobj.readline())
+
+        # Build header dictionary from the buffer.
+        hdr = dict(item.split(': ') for item in buf.rstrip().split('\n')[:-1])
+        hdr[Field.MAGIC_NUMBER] = magic_number
+
+        # Check integrity of TCK header.
+        if 'datatype' not in hdr:
+            raise HeaderError("Missing 'datatype' attribute in TCK header.")
+
+        if not hdr['datatype'].startswith('Float32'):
+            msg = ("TCK only supports float32 dtype but 'dataype: {}' was"
+                   " specified in the header.").format(hdr['datatype'])
+            raise HeaderError(msg)
+
+        if 'file' not in hdr:
```

Comment:
I would guess that's the best way to go, yes. Use it if there, else, use that heuristic.